Working update

holycrepe opened the pull request:
Summary of changes: 
1. Note validation (via external script)
2. Find deleted notes
3. Updated see also processing
4. Confirmed some old references to deleted functions were removed
5. README updated

Commit:
Added DTD Validation to makeNote/updateNote functions. These are for uploading content to Evernote's servers

Commit:
1. Move lxml code out of addon and into separate command-line script since unable to bundle lxml
2. Integrated command line utility with Anknotes' Anki menu

Commit:
Minor changes to logging

Commit:
Added Command Line Tools such as Find Deleted Notes

Commit:
Finalized Note validation and Note maintenance ('Find deleted notes')

Commit:
Class-based processing of See Also notes. This is to detect changes that need to be uploaded to Evernote

Commit:
Confirmed no references to generateTocTitle, all are to EvernoteNoteTitle.generateTOCTitle

Commit:
Squashed some bugs

Commit:
Long-overdue update for README, moved README to root, and added my /extra/dev folder to the repo

Commit:
Merge branch 'master' into holycrepe

Conflicts:
	.gitignore
	anknotes/Anki.py
	anknotes/AnkiNotePrototype.py
	anknotes/Controller.py
	anknotes/EvernoteImporter.py
	anknotes/EvernoteNoteFetcher.py
	anknotes/EvernoteNotePrototype.py
	anknotes/EvernoteNoteTitle.py
	anknotes/EvernoteNotes.py
	anknotes/__main__.py
	anknotes/ankEvernote.py
	anknotes/bare.py
	anknotes/constants.py
	anknotes/db.py
	anknotes/error.py
	anknotes/extra/ancillary/enml2.dtd
	anknotes/extra/ancillary/regex-see_also.txt
	anknotes/extra/ancillary/regex.txt
	anknotes/find_deleted_notes.py
	anknotes/graphics.py
	anknotes/html.py
	anknotes/logging.py
	anknotes/menu.py
	anknotes/settings.py
	anknotes/shared.py
	anknotes/structs.py
	anknotes/toc.py

Commit:
Merge branch 'holycrepe'

Commit:
Re-merge changes

Commit:
Fix api check on first load; Exclude developer automation files from repo

Commit:
Fix api check on first load; Exclude developer automation files from repo

Commit:
Merge branch 'master' into holycrepe

Commit:
Merge branch 'master' into brumar/holycrepe

Scriptkiddi committed:
fixed some startup problems

Scriptkiddi commented:
Hey still getting import errors i could fix the most of them but there
are a few NoneType and so on


On 23.09.2015 08:11, Avi Puchalapalli wrote:
> Summary of changes: 1. Note validation (via external script) 2.
> Find deleted notes 3. Updated see also processing 4. Confirmed some
> old references to deleted functions were removed 5. README updated
> 
> ------------------------------------------------------------------------
>
> 
> 
> You can view, comment on, or merge this pull request online at:
> 
> https://github.com/brumar/anknotes/pull/16
> 
> 
> Commit Summary
> 
> * Added DTD Validation to makeNote/updateNote functions. These are
> for uploading content to Evernote's servers * 1. Move lxml code out
> of addon and into separate command-line script since unable to
> bundle lxml * Minor changes to logging * Added Command Line Tools
> such as Find Deleted Notes * Finalized Note validation and Note
> maintenance ('Find deleted notes') * Class-based processing of See
> Also notes. This is to detect changes that need to be uploaded to
> Evernote * Confirmed no references to generateTocTitle, all are to 
> EvernoteNoteTitle.generateTOCTitle * Squashed some bugs *
> Long-overdue update for README, moved README to root, and added my 
> /extra/dev folder to the repo
> 
> 
> File Changes
> 
> * *M* .gitignore 
> <https://github.com/brumar/anknotes/pull/16/files#diff-0> (4) * *A*
> README.md <https://github.com/brumar/anknotes/pull/16/files#diff-1>
> (219) * *M* anknotes/Anki.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-2> (112) *
> *M* anknotes/AnkiNotePrototype.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-3> (296) *
> *M* anknotes/Controller.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-4> (333) *
> *M* anknotes/EvernoteImporter.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-5> (186) *
> *M* anknotes/EvernoteNoteFetcher.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-6> (75) *
> *M* anknotes/EvernoteNotePrototype.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-7> (30) *
> *M* anknotes/EvernoteNoteTitle.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-8> (100) *
> *M* anknotes/EvernoteNotes.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-9> (99) *
> *D* anknotes/README.md 
> <https://github.com/brumar/anknotes/pull/16/files#diff-10> (23) *
> *M* anknotes/__main__.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-11> (66) *
> *A* anknotes/_re.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-12> (277) *
> *M* anknotes/ankEvernote.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-13> (235) *
> *M* anknotes/bare.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-14> (347) *
> *M* anknotes/constants.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-15> (50) *
> *M* anknotes/db.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-16> (15) *
> *A* anknotes/enums.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-17> (108) *
> *M* anknotes/error.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-18> (4) *
> *A* anknotes/extra/ancillary/QMessageBox.css 
> <https://github.com/brumar/anknotes/pull/16/files#diff-19> (15) *
> *R* anknotes/extra/ancillary/enml2.dtd 
> <https://github.com/brumar/anknotes/pull/16/files#diff-20> (0) *
> *M* anknotes/extra/ancillary/regex-see_also.txt 
> <https://github.com/brumar/anknotes/pull/16/files#diff-21> (11) *
> *M* anknotes/extra/ancillary/regex.txt 
> <https://github.com/brumar/anknotes/pull/16/files#diff-22> (10) *
> *A* anknotes/extra/dev/Restart Anki.lnk 
> <https://github.com/brumar/anknotes/pull/16/files#diff-23> (0) *
> *A* anknotes/extra/dev/anknotes.developer 
> <https://github.com/brumar/anknotes/pull/16/files#diff-24> (0) *
> *A* anknotes/extra/dev/anknotes.developer.automate 
> <https://github.com/brumar/anknotes/pull/16/files#diff-25> (0) *
> *A* anknotes/extra/dev/anknotes_standAlone_template.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-26> (112) *
> *A* anknotes/extra/dev/anknotes_test.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-27> (209) *
> *A* anknotes/extra/dev/restart_anki.bat 
> <https://github.com/brumar/anknotes/pull/16/files#diff-28> (2) *
> *A* anknotes/extra/graphics/Tomato-icon.ico 
> <https://github.com/brumar/anknotes/pull/16/files#diff-29> (0) *
> *A* anknotes/extra/graphics/Tomato-icon.png 
> <https://github.com/brumar/anknotes/pull/16/files#diff-30> (0) *
> *M* anknotes/find_deleted_notes.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-31> (162) *
> *M* anknotes/graphics.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-32> (3) *
> *M* anknotes/html.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-33> (78) *
> *M* anknotes/logging.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-34> (459) *
> *M* anknotes/menu.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-35> (209) *
> *M* anknotes/settings.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-36> (12) *
> *M* anknotes/shared.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-37> (60) *
> *A* anknotes/stopwatch/__init__.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-38> (278) *
> *A* anknotes/stopwatch/tests/__init__.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-39> (71) *
> *M* anknotes/structs.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-40> (645) *
> *M* anknotes/toc.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-41> (76) *
> *A* anknotes_start_bare.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-42> (9) *
> *A* anknotes_start_find_deleted_notes.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-43> (11) *
> *A* anknotes_start_note_validation.py 
> <https://github.com/brumar/anknotes/pull/16/files#diff-44> (120)
> 
> 
> Patch Links:
> 
> * https://github.com/brumar/anknotes/pull/16.patch *
> https://github.com/brumar/anknotes/pull/16.diff
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/brumar/anknotes/pull/16>.
>

holycrepe commented:
Okay, thanks for looking into this. I'm running a code inspection with PyCharm now, and hopefully that will catch everything

Commit:
Merge remote-tracking branch 'origin/holycrepe' into holycrepe

Commit:
Rebuild template after obtaining Evernote uid, shard

Commit:
The Anki/Evernote imports actually need to be conditional, because this file is also used by the external command line utility to validate notes. Since that utility doesn't operate in the Anki environment, trying to import these modules will result in an import error. I changed the conditional to import if it's in Anki environment.

Commit:
Various Code Inspection fixes

holycrepe commented:
@Scriptkiddi Unfortunately PyCharm couldn't find anything that would cause an import error. Could you screenshot the errors and paste them here?

Scriptkiddi commented:
An error occurred in an add-on.
Please post on the add-on forum:
https://anki.tenderapp.com/discussions/add-ons

Traceback (most recent call last):
  File "/usr/share/anki/aqt/main.py", line 165, in onOpenProfile
    self.loadProfile()
  File "/usr/share/anki/aqt/main.py", line 244, in loadProfile
    runHook("profileLoaded")
  File "/usr/share/anki/anki/hooks.py", line 26, in runHook
    func(*args)
  File "/home/fritz/Anki/addons/anknotes/__main__.py", line 83, in anknotes_profile_loaded
    menu.upload_validated_notes(True)
  File "/home/fritz/Anki/addons/anknotes/menu.py", line 140, in upload_validated_notes
    controller = Controller()
  File "/home/fritz/Anki/addons/anknotes/Controller.py", line 50, in __init__
    self.evernote = Evernote()
  File "/home/fritz/Anki/addons/anknotes/ankEvernote.py", line 87, in __init__
    self.client = EvernoteClient(token=auth_token, sandbox=ANKNOTES.EVERNOTE_IS_SANDBOXED)
NameError: global name 'EvernoteClient' is not defined

Scriptkiddi commented:
i think i fixed this by moving the evernote client import out of your if statment

holycrepe commented:
Could you try this: http://pastebin.com/8111hStH

I added an assert statement at line 19. I also changed the conditional to test if Anki module is importable instead of lxml. It's possible that lxml is importable in your Anki config

holycrepe commented:
Unfortunately having it outside of the conditional will break the script I'm using to run note validation. I couldn't get lxml imported into my Anki config, so I made a script that runs externally via command line as opposed to within the Anki environment as an addon

holycrepe commented:
I fixed the issue with templates that was probably causing a string formatting error when you first loaded Anknotes. After your first import, Anknotes will now detect the uid and shard, and after next launch, will rebuild the templates. This is required for the template to be able to generate Evernote Note links from guids

Commit:
Templates will regenerate when new, valid shard/uid are detected

Scriptkiddi commented:
so the import errors are gone good job
now I tried to continue the test but when i open the settings i get 
An error occurred. It may have been caused by a harmless bug, 
or your deck may have a problem. 
To confirm it's not a problem with your deck, please run Tools > Check Database. 
If that doesn't fix the problem, please copy the following
into a bug report:
Traceback (most recent call last):
  File "/usr/share/anki/aqt/preferences.py", line 41, in reject
    self.accept()
  File "/usr/share/anki/aqt/preferences.py", line 37, in accept
    self.mw.reset()
  File "/usr/share/anki/aqt/main.py", line 433, in reset
    self.col.reset()
  File "/usr/share/anki/anki/collection.py", line 228, in reset
    self.sched.reset()
  File "/usr/share/anki/anki/sched.py", line 53, in reset
    self._resetLrn()
  File "/usr/share/anki/anki/sched.py", line 459, in _resetLrn
    self._resetLrnCount()
  File "/usr/share/anki/anki/sched.py", line 451, in _resetLrnCount
    self.dayCutoff) or 0
  File "/usr/share/anki/anki/db.py", line 72, in scalar
    res = self.execute(*a, **kw).fetchone()
TypeError: instance of cursor required for first argument

are you using the database cursor anywhere?

Scriptkiddi commented:
and if so for what?

holycrepe commented:
Hmm... I think I made a function that wraps around the cursor for easy access ... ankDB() in db.py. Does it still occur if you remove the addon from Anki? I don't think I do anything to the cursor. There are a few places where I have to call an external script so I close the collection entirely... But I don't think that's happening here.

I'm not sure honestly.... This happens when you open the preferences? can you add a log statement to
setup_evernote() method in line 69 of settings.py to debug the state of the cursor?

log_dump({'DB CON': mw.col.db._db, 'DB CUR': mw.col.db_db.cursor(), 'CUR DESC': mw.col.db._db.cursor().description})

This should dump data to extra\logs\dump.log

holycrepe commented:
Also try commenting out lines of anknotes_onload() method at line 102 of __main__.py and seeing if that resolves.

Scriptkiddi commented:
so if i take out anknotes it disappears also if i comment the lines in anknotes_onload() 
ok so i traced it back to upload_validated_notes

 [2015-09-23 19:57:25]:  **** Dumping Object
                              
                              {   'CUR DESC': None,
                                  'DB CON': <anki.db.DB object at 0x7fd03016df10>,
                                  'DB CUR': <pysqlite2.dbapi2.Cursor object at 0x7fd02275d6c0>}

oh and also could you please use explicit importing otherwise it is almost impossible to trace through your code

holycrepe commented:
By explicit importing do you mean not using the * in my import statements? If so sure I can do that.

It looks like the cursor is valid... I don't think the cursor description has to be set. The only modification happens at the end of that method - two executions and then a commit. I don't see how that would invalidate the cursor, however. 

Can you execute a query with the DB connection in settings.py? 

Also maybe try this change so the connection doesn't try to commit unless there is a query executed: https://gist.github.com/holycrepe/27bd14592e908eb6a776

holycrepe commented:
If you're unable to resolve the error by the end of the day, I'll set up a new Anki profile and see if that has something to do with the error

Commit:
Add excluded tags to settings and minor fixes

Scriptkiddi commented:
yes I mean from x import *

holycrepe commented:
Ok I will start replacing those over the next few days

Scriptkiddi commented:
so the move of the commit statment did not fix it
so i look a little bit deeper and found on line 130 in db.py (all) is calling self.execute with out a sql parameter i guess thats the missing cursor anki is complaning about

holycrepe commented:
Hmm good catch... But where is all() being called without a sql parameter? And why would the call to all() in upload_validated_notes result in a missing cursor?

The call is: `dbRows = ankDB().all("SELECT * FROM %s WHERE validation_status = 1 " % TABLES.MAKE_NOTE_QUEUE)`

Still a bit confused

holycrepe commented:
If you force a sql parameter for all() with the following code, do you get an Argument Error instead of the cursor error?

``    def all(self, sql, *a, **kw):``
``        return self.execute(sql, *a, **kw).fetchall()``

holycrepe commented:
Also the error you referenced is using the scalar method. Unless you think the call to .all() somehow messed up the cursor for the subsequent call to scalar by Anki. To save you some time, I changed all the methods to require a sql parameter:

https://gist.github.com/holycrepe/6d0456aaa4d4e57f6b41

Commit:
See Also tweak; Updated db.py to use forced sql parameters

Commit:
Init DB before adding models to Anki, fix note processing

holycrepe commented:
Unfortunately I couldn't reproduce the error, even with a new profile. Must have something to do with your particular setup

holycrepe commented:
Another step to try - Override the anki functions that are giving you an error so you can trace what's going on.

So if the scalar function is what's failing, you can try inserting the following code at the beginning of anknotes_onload: (I would expect that the self object is None, since the first argument to execute should be self. I don't think the failure is the sql argument, which would be second)


		def my_scalar(self, *a, **kw):
			 print [self, type(self)]
			 print [a, kw]
			 # more debug code here 

		mw.col.db.scalar = wrap(mw.col.db.scalar, my_scalar, "before")

holycrepe commented:
Also add the following to the wrap function so you can double check the path:
`` print self._db.path ``

There is a bug I just fixed where I hard coded the path for local db instances, but I don't think that's happening. To confirm, add an `assert False` or `raise` to  line 29 of db.py under `if dbLocal:`

I just went through every single line of code that affects the db, and I couldn't find any errors. Can you add assertions or raises to the following lines of menu.py, to ensure the profile is not being loaded or unloaded:

207 above `mw.unloadCollection()`, 
245 above `mw.loadCollection()`

Commit:
Support for multiple Anki profiles: User will need to fill ANKI_PROFILE_NAME under settings.

Commit:
Improvements to note template and CSS; Auto Update note models; better profile detection

Commit:
Add counter

Commit:
Auto reloading modules (dev function)
Automated progress counts and reporting

holycrepe commented:
@Scriptkiddi any luck with the errors? Can you try disabling the automated upload on start (set `ANKNOTES.AUTOMATE_VALIDATION` to `False` in `constants.py` or comment out the code in `__main__.py`?

Then, try an import or other database activity, then, try loading the preferences to see if any database activity is causing the error, or only the database activity associated with `upload_validated_notes()`? 

If you continue having issues I will set up Anki on a friends computer, and download a fresh copy of anknotes from my branch, and see if the issue is reproducible that way.

On a side note I completed, but did not push, updates to the 'See Also' links... They are fully functional now. Anknotes successfully uploaded notes with modified see also links to my Evernote account. So I think pretty much updates on my TODO list are done. It's a shame that we can't get it working on your setup, though.

@brumar, can you test and see if you are getting @Scriptkiddi's error?

holycrepe committed:
Disabled remove_tags scripts in Anki environment

Commit:
Fix note store check on import

Commit:
Merge branch 'brumar/holycrepe' into holycrepe

Commit:
Merge branch 'holycrepe' of github.com:brumar/anknotes into holycrepe

Commit:
Bug fixes that would affect new users

Commit:
Various bug fixes and improvements

Scriptkiddi commented:
hey i just spend the weekend in munich so i will get to it maybe tomorrow

brumar commented:
I will have a look on it tonight.

Commit:
Bug fix, update tag deletion check

Commit:
Bug fix, update tag deletion check

Commit:
Merge branches 'holycrepe' and 'holycrepe' of github.com:brumar/anknotes into holycrepe

# Conflicts:
#	anknotes/ankEvernote.py
#	anknotes/shared.py

Commit:
Fix unicode error with title check

Commit:
Update last commit; fix local line endings

Commit:
Update title check / counters inconsistency

holycrepe commented:
Thanks guys. I setup a new account and debugged everything from scratch so the core functionality should be good to go. There are some issues with automation (e.g. using the `"Complete All Steps"` function in `Process See Also Footer Links` menu. I think I need to put more checks to ensure the clients are properly loaded and db is not locked since we have to close out the Anki profile and run an external script. 

So, feel free to ignore that functionality when doing your testing.

Also - dev functions: there's some info in the readme, but in addition, I added the ability to automatically reload the main `anknotes.Controller` module every time you click a menu item, which also forces reload of the `anknotes.Evernote` and `anknotes.Anki` classes. This is purely for testing, when making multiple changes to these classes, so you don't have to keep quitting and restarting Anki to test your changes. (Side note: I included a batch file that silently reloads Anki in the `/extra/dev/` folder). Finally, for testing purposes, you can add code to main.py that automatically executes every time the Anki profile is loaded, which is useful if you're updating something other than the `Controller`, `Evernote`, or `Anki` classes (since those are the only ones that automatically reload when you click a menu item). 

For these dev functions, you'll have to create empty files `anknotes.developer` and `anknotes.developer.automate` in the `/extra/dev/` folder.

Commit:
Ensure TOC enex exists in find_deleted_notes

Commit:
Created example user configuration file (constants_user.py)

Commit:
Fix unicode/string encoding issues

Commit:
Major improvements to Unicode error handling

Commit:
Safer deletion of log folders

Commit:
Fix for user constants file

Commit:
Bug fix for find deleted notes

Commit:
Fix argument processing

Commit:
Minor changes

Scriptkiddi committed:
replaced tabs with 4 spaces and fixed some case where a emtpy evernote guid set would cause crashes

Scriptkiddi commented:
so first of please follow https://www.python.org/dev/peps/pep-0008/
makes the code a lot more readable and please do not use one line if statments they are really hard to read when going over the code real quick

also my database error is still there
An error occurred. It may have been caused by a harmless bug, 
or your deck may have a problem. 
To confirm it's not a problem with your deck, please run Tools > Check Database. 
If that doesn't fix the problem, please copy the following
into a bug report:
Traceback (most recent call last):
  File "/usr/share/anki/aqt/preferences.py", line 41, in reject
    self.accept()
  File "/usr/share/anki/aqt/preferences.py", line 37, in accept
    self.mw.reset()
  File "/usr/share/anki/aqt/main.py", line 433, in reset
    self.col.reset()
  File "/usr/share/anki/anki/collection.py", line 228, in reset
    self.sched.reset()
  File "/usr/share/anki/anki/sched.py", line 53, in reset
    self._resetLrn()
  File "/usr/share/anki/anki/sched.py", line 459, in _resetLrn
    self._resetLrnCount()
  File "/usr/share/anki/anki/sched.py", line 451, in _resetLrnCount
    self.dayCutoff) or 0
  File "/usr/share/anki/anki/db.py", line 72, in scalar
    res = self.execute(*a, **kw).fetchone()
TypeError: instance of cursor required for first argument

Commit:
Search Improvements, See Also fixes

holycrepe commented:
Sorry for the late response, I didn't see a notification from github. I'll make the formatting changes you requested, and avoid the one line statements. I'll run a code formatting tool to fix this

Were you able to further debug the database error? E.g. my suggestion to override Anki's db method:

		def my_scalar(self, *a, **kw):
			 print [self, type(self)]
			 print [a, kw]
			 # more debug code here 

		mw.col.db.scalar = wrap(mw.col.db.scalar, my_scalar, "before")

Does disabling script validation help?

Commit:
Merge branches 'holycrepe' and 'holycrepe' of github.com:brumar/anknotes into holycrepe

# Conflicts:
#	anknotes/Anki.py
#	anknotes/AnkiNotePrototype.py
#	anknotes/Controller.py
#	anknotes/EvernoteNoteFetcher.py
#	anknotes/EvernoteNotes.py
#	anknotes/__main__.py
#	anknotes/ankEvernote.py
#	anknotes/constants.py
#	anknotes/constants_user.py
#	anknotes/counters.py
#	anknotes/db.py
#	anknotes/detect_see_also_changes.py
#	anknotes/find_deleted_notes.py
#	anknotes/html.py
#	anknotes/logging.py
#	anknotes/menu.py
#	anknotes/settings.py
#	anknotes/shared.py
#	anknotes/structs.py

holycrepe commented:
Also if you could log the sql from all queries, and figure out the last query executed before the cursor issues, that would be very helpful

Commit:
Attempt to remerge changes

holycrepe commented:
I had to revert your latest commit because I couldn't get it to merge properly and I made several major changes to the search algorithms. I'm guessing you just ran a code formatting tool and then fixed the bug you mentioned in your commit message. 

Can you please get the latest commit, re-fix the bug you caught, and then make a commit? That way I can see what exactly the bug was and the diff for the bug doesn't get lost on all the formatting changes. Then feel free to re-run your formatting tool. From here on out, I will keep the code consistent with the formatting style you referenced in your link

I am in the process of changing the implicit imports to explicit like you requested, but that may take some time. In the meantime please consider using an IDE that will help you navigate the code. For example PyCharm will navigate properly even with implicit imports, and plugins using ctags (e.g. SourceCookifier for Notepad++) will also navigate the code properly.

Commit:
Package sqlite3.dll in /extra/ancillary folder

holycrepe commented:
Also please try overwriting sqlite3.dll in the Anki program dir (e.g. C:\Program Files (x86)\Anki) with the latest dll, located in /extra/ancillary/ as of latest commit

Commit:
Debug database error

Commit:
Exclude constants_user.py from repo

Commit:
Debug database errors

holycrepe commented:
Hey @Scriptkiddi , I created debug wrappers for anki's db methods. Please test the latest commit, and send me a zip of your `/extra/logs/sql folder`

holycrepe commented:
Also, if you have errors with the latest commit, save the following file as `constants_user.py` to the `anknotes` folder alongside `constants.py` and re-test (It will disable uploading and automated validation)

http://pastebin.com/FDg3NKw1

Commit:
Create subnotes: Filter headings

Commit:
Convert one line if/for/try statements to two lines

Commit:
Minor bug fixes

holycrepe commented:
I converted all one line statements to two lines. Hope it improves the readability. The rest of the formatting changes and explicit imports are in progress

Commit:
PEP 0008 Changes, timer wrapper

Commit:
Database, Logging improvements

Commit:
Change gitignore

Commit:
Changes to create_subnotes

Commit:
Update TOC Name Priority, remove trailing spaces

Commit:
Subnotes automation and fixes

Commit:
Logging improvements

Commit:
Update create_subnotes logging

Commit:
Headings: Ignore plurality

Scriptkiddi commented:
so sorry for the long wait I was really busy first of im using pycharm still and i will get to the testing later today

Commit:
Don't add trailing newline to file that tracks last Anki profile name

Scriptkiddi commented:
launching the newst version on windows 10 I get this error

An error occurred in an add-on.
Please post on the add-on forum:
https://anki.tenderapp.com/discussions/add-ons

Traceback (most recent call last):
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\__main__.py", line 308, in main
    controller = Controller()
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\__main__.py", line 257, in __init__
    self.anki.add_evernote_model()
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\__main__.py", line 128, in add_evernote_model
    source_ord, source_field = fmap[GUID_FIELD_NAME]
KeyError: 'Evernote GUID'

Commit:
Fix error with EvernoteNotes.addDBNote when baseQuery is blank

holycrepe commented:
Oh sorry ... Seems to be an issue with initializing the templates. I'll fix that now

holycrepe commented:
@Scriptkiddi, Actually your Windows 10 error seems really odd to me. It looks like one of your files is out of date since:

1. The self.anki.add_evernote_model call has been moved to Controller.py 
2. The constant 'GUID_FIELD_NAME' has been renamed to FIELDS.EVERNOTE_GUID
3. A search of current source code yields no results for 'source_ord' or 'source_field'

Since the self.anki.add_evernote_model call is in `__main__.py` for you, and the `GUID_FIELD_NAME` variable resolves, I think your `__main__.py` and `constants.py` are out of date. 

If you have symlinks or something like that set up, maybe they are pointing to an old copy of the code. Can you confirm that they are up to date by opening anknotes\__main__.py in a text editor? Can you try re-downloading or re-cloning the holycrepe branch in this repository?

Commit:
Update in_anki check

Commit:
Simplify creation of elements for settings panel

Scriptkiddi commented:
my bad i downloaded the wrong branch i will test it later today again

Scriptkiddi commented:
An error occurred in an add-on.
Please post on the add-on forum:
https://anki.tenderapp.com/discussions/add-ons

Traceback (most recent call last):
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\menu.py", line 72, in <lambda>
    def auto_reload_wrapper(function): return lambda: auto_reload_modules(function)
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\menu.py", line 89, in auto_reload_modules
    function()
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\menu.py", line 156, in import_from_evernote
    controller.proceed()
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\Controller.py", line 210, in proceed
    self.evernoteImporter.proceed(auto_paging)
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\EvernoteImporter.py", line 176, in proceed
    self.proceed_import_notes()
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\EvernoteImporter.py", line 235, in proceed_import_notes
    self.ImportProgress.processResults(self.import_into_anki(self.ImportProgress.GUIDs.Server.New))
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\EvernoteImporter.py", line 132, in import_into_anki
    Results = self.evernote.create_evernote_notes(evernote_guids)
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\ankEvernote.py", line 370, in create_evernote_notes
    self.check_ancillary_data_up_to_date()
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\ankEvernote.py", line 397, in check_ancillary_data_up_to_date
    "Tags were not up to date when checking ancillary data")
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\ankEvernote.py", line 537, in update_tags_database
    db.executemany(enTag.sqlUpdateQuery(), data)
  File "C:\Users\Scriptkiddi\Documents\Anki\addons\anknotes\db.py", line 441, in executemany
    self._db.executemany(sql, data)
sqlite3.OperationalError: no such table: anknotes_evernote_tags

holycrepe commented:
Okay looks like the tags db was not created on launch. I will fix that asap

Commit:
Automate generating and storing settings keys plus bug fixes

Commit:
Clean code

Commit:
Filter logs

holycrepe commented:
Sorry for the delay... Bug is fixed with the latest commit. Please test the latest commit, and if you still get the database errors, send me a zip of your `/extra/logs` folder. Thanks for your efforts

Commit:
Code improvements

Scriptkiddi commented:
hey so i had some time to look at it a little bit longer i fixed to import errors but my time ran out on 
An error occurred in an add-on.
Please post on the add-on forum:
https://anki.tenderapp.com/discussions/add-ons

Traceback (most recent call last):
  File "/usr/share/anki/aqt/addons.py", line 41, in loadAddons
    __import__(file.replace(".py", ""))
  File "/home/fritz/Anki/addons/anknotes_start.py", line 6, in <module>
    from anknotes import __main__
  File "/home/fritz/Anki/addons/anknotes/__main__.py", line 17, in <module>
    from anknotes.shared import *
  File "/home/fritz/Anki/addons/anknotes/shared.py", line 14, in <module>
    from anknotes.constants import *
  File "/home/fritz/Anki/addons/anknotes/constants.py", line 4, in <module>
    from anknotes.constants_settings import *
  File "/home/fritz/Anki/addons/anknotes/constants_settings.py", line 12, in <module>
    SETTINGS = DictSettings(key='anknotes')
  File "/home/fritz/Anki/addons/anknotes/dicts.py", line 92, in __init__
    super(cls.mro()[mro], self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/dicts.py", line 79, in __init__
    super(cls_mro, self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/dicts.py", line 11, in __init__
    super(cls.mro()[mro], self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/dicts_base.py", line 123, in __init__
    super(cls.mro()[mro], self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/addict/addict.py", line 46, in __init__
    self.update(*a, **kw)
  File "/home/fritz/Anki/addons/anknotes/addict/addict.py", line 366, in update
    "{0} types".format(arg.__class__))
TypeError: Dict does not understand <type 'int'> types
Traceback (most recent call last):
  File "/usr/share/anki/aqt/addons.py", line 41, in loadAddons
    __import__(file.replace(".py", ""))
  File "/home/fritz/Anki/addons/anknotes_start_note_validation.py", line 2, in <module>
    from anknotes import stopwatch
  File "/home/fritz/Anki/addons/anknotes/stopwatch/__init__.py", line 12, in <module>
    from anknotes.constants import ANKNOTES
  File "/home/fritz/Anki/addons/anknotes/constants.py", line 4, in <module>
    from anknotes.constants_settings import *
  File "/home/fritz/Anki/addons/anknotes/constants_settings.py", line 12, in <module>
    SETTINGS = DictSettings(key='anknotes')
  File "/home/fritz/Anki/addons/anknotes/dicts.py", line 92, in __init__
    super(cls.mro()[mro], self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/dicts.py", line 79, in __init__
    super(cls_mro, self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/dicts.py", line 11, in __init__
    super(cls.mro()[mro], self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/dicts_base.py", line 123, in __init__
    super(cls.mro()[mro], self).__init__(mro+1, *a, **kw)
  File "/home/fritz/Anki/addons/anknotes/addict/addict.py", line 46, in __init__
    self.update(*a, **kw)
  File "/home/fritz/Anki/addons/anknotes/addict/addict.py", line 366, in update
    "{0} types".format(arg.__class__))
TypeError: Dict does not understand <type 'int'> types

since the error logs did not log anything new im not attaching them

Scriptkiddi committed:
import error in addict.py and tuple index deletion problem in __get__args in DictString

Scriptkiddi commented:
oh also please keep all external libaries in the anknotes folder since its a really bad practice installing the own deps into folders where they dont belong

holycrepe commented:
Thanks for looking into it! Now that I have some experience with Python (I never coded in Python before this project), I am in the process of trying to remove all the crappy code that has made it into my branch in the process of trying to learn the language. So in the next few days I should have a build without most of the problems with the current branch (For example, using a nested dictionary for the settings object, while it saved time when creating new settings, seems like unnecessary overhead). 

Do you mean to put the dependencies in the root folder of the repo (along with `anknotes_start.py`) or the `/anknotes` folder along with `__main__.py`?

Scriptkiddi commented:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

I mean putting the bs4 lib in the anknotes folder you can not move the
anknotes_start in the anknotes folder because then anki would not load
the plugin

On 16.11.2015 23:22, Avi Puchalapalli wrote:
> Thanks for looking into it! Now that I have some experience with
> Python (I never coded in Python before this project), I am in the
> process of trying to remove all the crappy code that has made it
> into my branch in the process of trying to learn the language. So
> in the next few days I should have a build without most of the
> problems with the current branch (For example, using a nested
> dictionary for the settings object, while it saved time when
> creating new settings, seems like unnecessary overhead).
> 
> Do you mean to put the dependencies in the root folder of the repo 
> (along with |anknotes_start.py|) or the |/anknotes| folder along
> with |__main__.py|?
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/brumar/anknotes/pull/16#issuecomment-157190072>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQEcBAEBCAAGBQJWSlddAAoJEMlXFywT0VZDNqYH/RHT0TytASDKb8OVqoz9+FZ9
yLYI48n3uPyC434quw6v+sq+UnoHL+5ChhJ3KCMU1NgurtGVcD4TmWtwROoMLeF5
yGoSh5imauCEORnsRVimriGSdrCfqq9zXw7PY0lRmDI9CBvka4uijYELhTVGuT5L
EKVCQljuYO4X0TkvxR2wGN1XyDMVx4VJHgtd1qSUOCNGFmV5u4UqFMKCmrmwCwaD
F2+smduov8bN+ZfpO1HMusPhnE1Eh/PSUGmBy1n+HX7S+DZ437pzCClZ9IjXEtdy
HoXKNPVxLrf1D+gsnqPNEFndAN5v4jiAo0hItyc6mcE7FnSt6vmujh7KlRCkvYc=
=G9hY
-----END PGP SIGNATURE-----

holycrepe commented:
Oh okay so move bs4 from the repository root to the same folder as `anknotes/__main__.py`? The only issue with that is that the imports in bs4 and other libraries all assume that it is a top-level package; So I would have to prefix all the imports in bs4 with 'anknotes.' so that it is properly routed to the anknotes folder. The other option would be converting bs4's imports from absolute to relative

This is fairly simple for bs4 since it only has a few import statements, but I can see how it can get complicated. I see that the evernote, oauth2, and thrift libraries' imports are all prefixed with `anknotes.`. Did that have to be done manually or is there a tool that will do that for us?

I already updated the imports for bs4 manually as you recommended, but I'm just curious for future reference

Scriptkiddi commented:
i did it manually but im pretty sure there are tools for that kind of stuff

sjobeek commented:
Hey guys, this is some interesting work.  

I'm a python dev and looking to potentially use this for my own study and of course contribute back fixes I may need to make.

Which branch would you suggest I start with, and do you have any tips or obvious pitfalls I should avoid when trying to get this up and running on my system for the first time?

Scriptkiddi commented:
Hey I would suggest you either start from holycrepes branch but there I dont know how far he is, I have not touched the project in a while or you take the master branch which is far behind but relaltivly stable. I would recommend skimming through the evernote api guide lines and geoi tting familiar how anki works internally (decks, notes, cards).

holycrepe commented:
@sjobeek My branch has some improvements but I'm not sure about its stability or whether I have pushed all local changes. I have an important exam coming up next week, but I'll be happy to look into this for you the week after.  I updated the `readme` on my repository so that should give you a good start.

In the meantime, follow @Scriptkiddi's suggestions. If you can try to reproduce the bug @Scriptkiddi mentioned [above](https://github.com/brumar/anknotes/pull/16#issuecomment-148987494), that would be very helpful

brumar commented:
Hello and welcome. I am bit out of sync with this project, I can't be very helpful at the moment. As a matter of practical use of the plug-in, I tried to find you the last version of the plugin I uploaded on anki/addons. But I can't find it anymore. Very probably because my account has expired...

KrishaanKhubchand commented:
hey guys, I've been trying to install this - But to no avail, any suggestions?

holycrepe commented:
Are you having issues with this PR or `brumar/master`?

brumar commented:
Hi Holycrepe !
I hope you are doing well.
I fixed the problem on my repo and added a token protection feature. @KrishaanKhubchand  and I talked by private message and now it seems to work for him. I also re-upload the addon here https://ankiweb.net/shared/info/999519295 . I also tried again to test your fork but with no success :(. By the way, are you the uploader of this one https://ankiweb.net/shared/info/1619377479 ? This is just a zip version of your repository, so that can't possibly work. 
Cheers !

holycrepe commented:
Hey Bruno, my AnkiWeb account expired `04/2016` so the upload is not mine. I'm glad you've been able to revive the project. I've had to reformat since working on anknotes, so I don't have a way to work on development right now.  If I can restore the repo from a backup, I may give it a shot.

Is there a specific issue with the fork? I was unable to reproduce any errors, and the fork was working as intended last time I used it, although it has been a long time.

brumar commented:
I got the same problem of account expiration which discouraged me a bit, because at that time I made a nice user documentation on the addon page (but without backup, shame on me). If you want to work again on this project I can retry to see what kind of specific error I got.

holycrepe commented:
Yes please, I would appreciate that. Guess we'll have to add the addon description to the repo so it's under version control! I have some exams coming up and it will be helpful to get my fork working. I'll be reinstalling anki, so maybe I'll be able to reproduce the error with a fresh environment and machine

brumar commented:
> Guess we'll have to add the addon description to the repo so it's under version control!
My thoughts exactly ahah : https://github.com/brumar/anknotes/blob/master/addonDescription.html

holycrepe commented:
Ah nice, I've committed a quick fix to that description: https://github.com/brumar/anknotes/commit/1ab3fedd4c4c9529b32e12ce4ac6dc9574c6d1f6
